x-pack/filebeat/input/httpjson/requester.go

```diff
@@ -0,0 +1,265 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package httpjson
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/pkg/errors"
+
+	stateless "github.com/elastic/beats/v7/filebeat/input/v2/input-stateless"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+type requestInfo struct {
+	url        string
+	contentMap common.MapStr
+	headers    common.MapStr
+}
+
+type requester struct {
+	log         *logp.Logger
+	client      *http.Client
+	dateCursor  *dateCursor
+	rateLimiter *rateLimiter
+	pagination  *pagination
+
+	method        string
+	reqBody       common.MapStr
+	headers       common.MapStr
+	noHTTPBody    bool
+	apiKey        string
+	authScheme    string
+	jsonObjects   string
+	splitEventsBy string
+}
+
+func newRequester(
+	config config,
+	rateLimiter *rateLimiter,
+	dateCursor *dateCursor,
+	pagination *pagination,
+	client *http.Client,
+	log *logp.Logger) *requester {
+	return &requester{
+		log:           log,
+		client:        client,
+		rateLimiter:   rateLimiter,
+		dateCursor:    dateCursor,
+		pagination:    pagination,
+		method:        config.HTTPMethod,
+		reqBody:       config.HTTPRequestBody.Clone(),
+		headers:       config.HTTPHeaders.Clone(),
+		noHTTPBody:    config.NoHTTPBody,
+		apiKey:        config.APIKey,
+		authScheme:    config.AuthenticationScheme,
+		splitEventsBy: config.SplitEventsBy,
+		jsonObjects:   config.JSONObjects,
+	}
+}
+
+type response struct {
+	header http.Header
+	body   common.MapStr
+}
+
+// processHTTPRequest processes HTTP request, and handles pagination if enabled
+func (r *requester) processHTTPRequest(ctx context.Context, publisher stateless.Publisher) error {
+	ri := &requestInfo{
+		url:        r.dateCursor.getURL(),
+		contentMap: common.MapStr{},
+		headers:    r.headers,
+	}
+
+	if r.method == "POST" && r.reqBody != nil {
+		ri.contentMap.Update(common.MapStr(r.reqBody))
+	}
+
+	var (
+		m, v     interface{}
+		response response
+		lastObj  common.MapStr
+	)
+
+	// always request at least once
+	hasNext := true
+
+	for hasNext {
+		resp, err := r.rateLimiter.execute(
+			ctx,
+			func(ctx context.Context) (*http.Response, error) {
+				req, err := r.createHTTPRequest(ctx, ri)
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed to create http request")
+				}
+				msg, err := r.client.Do(req)
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed to execute http client.Do")
```

Comment:
It looks like this error is passed through to `Run`. If so the input shuts down after an IO error. Timeouts/disconnects or other network errors should not shut down the input. These are recoverable errors in a sense of "let's try again".

errors.Wrap and similar make it difficult to "judge" the error (`errors.Is` can help to check for a cause). Consider to either log recoverable errors or create special error types that allow you to see what's going on.

About `errors.Wrap`: We should stop using the pkg/errors package. It is discontinued. At the same time go 1.14 introduced something similar via `fmt.Errorf` with `%w`. Better use `fmt.Errorf("failed to execute http request: %w", err)`